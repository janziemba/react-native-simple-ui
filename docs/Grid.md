# Grid
## Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props

## Column
### Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props
- `size` (_number_,  default null)

## Row
### Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props
- `size` (_number_, default null)

## Examples

```jsx
<Grid>
    <Column>
        <Text>
            Column
        </Text>
    </Column>
    <Column>
        <Text>
            Column
        </Text>
    </Column>
</Grid>
```
