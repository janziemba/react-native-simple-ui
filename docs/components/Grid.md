---
id: grid
title: Grid
---
## Grid
### Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props

## Column
### Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props
- `size` (_number_,  default null)

## Row
### Props
Extend React Native's [View](https://facebook.github.io/react-native/docs/view#props) props
- `size` (_number_, default null)

<img src="../assets/Grid/1.png" alt="Grid" width="250x" style="display: inline-block;">

## Examples

```jsx
<Grid>
    <Column>
        <Text>
            Column
        </Text>
    </Column>
    <Column>
        <Text>
            Column
        </Text>
    </Column>
</Grid>
```
