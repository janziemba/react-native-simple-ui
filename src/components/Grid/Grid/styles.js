export default () => ({
    base: {
        flexShrink: 1,
    },
    directions: {
        column: {
            flexDirection: 'column',
        },
        row: {
            flexDirection: 'row',
        },
    },
});
