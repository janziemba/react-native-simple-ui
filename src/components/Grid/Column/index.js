// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { StyleSheet, View } from 'react-native';

import type { ViewProps } from 'react-native/Libraries/Components/View/ViewPropTypes';

import withTheme from '../../../themes/withTheme';
import hasStyleChanged from '../../../utils/hasStyleChanged';
import styles from './styles';
import type { StylesType } from './styles';

type Props = ViewProps & {
    size?: number,
    styles: StylesType,
};

const defaultProps = {
    ...View.defaultProps,
    size: null,
};

type State = {
    styles: Array<StyleSheet.Styles>,
};

const getStyles = (props: Props): Array<StyleSheet.Styles> => {
    const { size, styles } = props;

    let flex = 1;

    if (size) {
        flex = size;
    } else if (styles.base.width) {
        flex = 0;
    }

    return [
        styles.base,
        { flex },
    ];
};

class Column extends PureComponent<Props, State> {
    static defaultProps = defaultProps;
    state = {
        styles: getStyles(this.props),
    };
    componentWillReceiveProps(nextProps): void {
        const propsOnWhichDependsTheStyle: Array<string> = ['size'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState({ styles: getStyles(nextProps) });
        }
    }
    render(): Node {
        const { children } = this.props;
        const { styles } = this.state;

        return (
            <View
                {...this.props}
                style={styles}
            >
                {children}
            </View>
        );
    }
}

Column = withTheme(styles, 'Column')(Column);

export default Column;
