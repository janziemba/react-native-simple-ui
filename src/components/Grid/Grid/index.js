// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { StyleSheet, View } from 'react-native';

import type { ViewProps } from 'react-native/Libraries/Components/View/ViewPropTypes';

import withTheme from '../../../themes/withTheme';
import hasStyleChanged from '../../../utils/hasStyleChanged';
import styles from './styles';
import type { StylesType } from './styles';

type Props = ViewProps & {
    children: Node,
    styles: StylesType,
};

const defaultProps = {
    ...View.defaultProps,
};

type State = {
    styles: Array<StyleSheet.Styles>,
};

const getStyles = (props: Props): Array<StyleSheet.Styles> => {
    const { children, styles } = props;

    const components = children.length > 1 ? children : [children];

    const isDirectionRow: boolean = components.find(child => child.type.originalComponentName === 'Row') !== undefined;

    return [
        styles.base,
        styles.directions[isDirectionRow ? 'column' : 'row'],
    ];
};

class Grid extends PureComponent<Props, State> {
    static defaultProps = defaultProps;
    state = {
        styles: getStyles(this.props),
    };
    componentWillReceiveProps(nextProps): void {
        const propsOnWhichDependsTheStyle: Array<string> = ['children'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState({ styles: getStyles(nextProps) });
        }
    }
    render(): Node {
        const { children } = this.props;
        const { styles } = this.state;

        return (
            <View
                {...this.props}
                style={styles}
            >
                {children}
            </View>
        );
    }
}

Grid = withTheme(styles, 'Grid')(Grid);

export default Grid;
