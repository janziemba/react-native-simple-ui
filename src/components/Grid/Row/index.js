// @flow

import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { StyleSheet, View } from 'react-native';

import type { ViewProps } from 'react-native/Libraries/Components/View/ViewPropTypes';

import withTheme from '../../../themes/withTheme';
import hasStyleChanged from '../../../utils/hasStyleChanged';
import styles from './styles';
import type { StylesType } from './styles';

type Props = ViewProps & {
    size?: number,
    styles: StylesType,
};

const defaultProps = {
    ...View.defaultProps,
    size: null,
};

type State = {
    styles: Array<StyleSheet.Styles>,
};

const getStyles = (props: Props): Array<StyleSheet.Styles> => {
    const { size, style, styles } = props;

    let flex = 1;

    if (size) {
        flex = size;
    } else if (style && style.width) {
        flex = 0;
    }

    return [
        style,
        styles.base,
        { flex },
    ];
};

class Row extends PureComponent<Props, State> {
    static defaultProps = defaultProps;
    state = {
        styles: getStyles(this.props),
    };
    componentWillReceiveProps(nextProps: Props): void {
        const propsOnWhichDependsTheStyle: Array<string> = ['size', 'style'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState({ styles: getStyles(nextProps) });
        }
    }
    render(): Node {
        const { children } = this.props;
        const { styles } = this.state;

        return (
            <View
                {...this.props}
                style={styles}
            >
                {children}
            </View>
        );
    }
}

Row = withTheme(styles, 'Row')(Row);

export default Row;
