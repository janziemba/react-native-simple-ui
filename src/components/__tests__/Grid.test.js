import React from 'react';
import renderer from 'react-test-renderer';

import { Column, Grid, Row } from '../Grid';

it('Column renders correctly', () => {
    const tree = renderer.create(<Column />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Row renders correctly', () => {
    const tree = renderer.create(<Row />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Grid with Column renders correctly', () => {
    const tree = renderer.create(
        <Grid>
            <Column />
        </Grid>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('Grid with Row renders correctly', () => {
    const tree = renderer.create(
        <Grid>
            <Row />
        </Grid>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
