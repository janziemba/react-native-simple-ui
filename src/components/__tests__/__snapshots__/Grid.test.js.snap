// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column renders correctly 1`] = `
<View
  size={null}
  style={
    Array [
      undefined,
      Object {
        "flexDirection": "column",
      },
      Object {
        "flex": 1,
      },
    ]
  }
  styles={
    Object {
      "base": Object {
        "flexDirection": "column",
      },
    }
  }
  theme={
    Object {
      "colors": Object {
        "layout": Object {
          "backgroundColor": "#ffffff",
        },
        "palette": Object {
          "danger": "#ff4136",
          "dark": "#484848",
          "disabled": "#999999",
          "info": "#75caeb",
          "light": "#f6f6f6",
          "primary": "#158cba",
          "success": "#28b62c",
          "warning": "#ff851b",
        },
        "social": Object {
          "facebook": "#3b5998",
          "google": "#dc4e41",
          "instagram": "#c32aa3",
          "linkedin": "#0077b5",
          "twitter": "#55acee",
          "vk": "#4c75a3",
        },
        "text": Object {
          "disabled": "#999999",
          "placeholder": "#e2e4e6",
          "primary": "#3d3d3d",
          "secondary": "#9e9e9e",
        },
      },
      "components": Object {},
      "radiuses": Object {
        "gigantic": 16,
        "huge": 12,
        "large": 8,
        "medium": 6,
        "small": 4,
        "tiny": 2,
      },
      "spacing": Object {
        "gigantic": 24,
        "huge": 20,
        "large": 16,
        "medium": 12,
        "small": 8,
        "tiny": 4,
      },
      "typography": Object {
        "fontFamily": "System",
        "fontSizes": Object {
          "large": 16,
          "medium": 14,
          "small": 12,
        },
        "fontWeights": Object {
          "bold": "900",
          "medium": "800",
          "normal": "600",
        },
        "iconSet": "MaterialIcons",
        "sizes": Object {
          "h1": Object {
            "fontSize": 31,
            "fontWeight": "900",
          },
          "h2": Object {
            "fontSize": 25,
            "fontWeight": "900",
          },
          "h3": Object {
            "fontSize": 19,
            "fontWeight": "800",
          },
          "h4": Object {
            "fontSize": 17,
            "fontWeight": "800",
          },
          "h5": Object {
            "fontSize": 16,
            "fontWeight": "800",
          },
          "h6": Object {
            "fontSize": 15,
            "fontWeight": "700",
          },
          "normal": Object {
            "fontSize": 14,
            "fontWeight": "700",
          },
          "small": Object {
            "fontSize": 12,
          },
        },
      },
    }
  }
/>
`;

exports[`Grid with Column renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "flexShrink": 1,
      },
      Object {
        "flexDirection": "row",
      },
    ]
  }
  styles={
    Object {
      "base": Object {
        "flexShrink": 1,
      },
      "directions": Object {
        "column": Object {
          "flexDirection": "column",
        },
        "row": Object {
          "flexDirection": "row",
        },
      },
    }
  }
  theme={
    Object {
      "colors": Object {
        "layout": Object {
          "backgroundColor": "#ffffff",
        },
        "palette": Object {
          "danger": "#ff4136",
          "dark": "#484848",
          "disabled": "#999999",
          "info": "#75caeb",
          "light": "#f6f6f6",
          "primary": "#158cba",
          "success": "#28b62c",
          "warning": "#ff851b",
        },
        "social": Object {
          "facebook": "#3b5998",
          "google": "#dc4e41",
          "instagram": "#c32aa3",
          "linkedin": "#0077b5",
          "twitter": "#55acee",
          "vk": "#4c75a3",
        },
        "text": Object {
          "disabled": "#999999",
          "placeholder": "#e2e4e6",
          "primary": "#3d3d3d",
          "secondary": "#9e9e9e",
        },
      },
      "components": Object {},
      "radiuses": Object {
        "gigantic": 16,
        "huge": 12,
        "large": 8,
        "medium": 6,
        "small": 4,
        "tiny": 2,
      },
      "spacing": Object {
        "gigantic": 24,
        "huge": 20,
        "large": 16,
        "medium": 12,
        "small": 8,
        "tiny": 4,
      },
      "typography": Object {
        "fontFamily": "System",
        "fontSizes": Object {
          "large": 16,
          "medium": 14,
          "small": 12,
        },
        "fontWeights": Object {
          "bold": "900",
          "medium": "800",
          "normal": "600",
        },
        "iconSet": "MaterialIcons",
        "sizes": Object {
          "h1": Object {
            "fontSize": 31,
            "fontWeight": "900",
          },
          "h2": Object {
            "fontSize": 25,
            "fontWeight": "900",
          },
          "h3": Object {
            "fontSize": 19,
            "fontWeight": "800",
          },
          "h4": Object {
            "fontSize": 17,
            "fontWeight": "800",
          },
          "h5": Object {
            "fontSize": 16,
            "fontWeight": "800",
          },
          "h6": Object {
            "fontSize": 15,
            "fontWeight": "700",
          },
          "normal": Object {
            "fontSize": 14,
            "fontWeight": "700",
          },
          "small": Object {
            "fontSize": 12,
          },
        },
      },
    }
  }
>
  <View
    size={null}
    style={
      Array [
        undefined,
        Object {
          "flexDirection": "column",
        },
        Object {
          "flex": 1,
        },
      ]
    }
    styles={
      Object {
        "base": Object {
          "flexDirection": "column",
        },
      }
    }
    theme={
      Object {
        "colors": Object {
          "layout": Object {
            "backgroundColor": "#ffffff",
          },
          "palette": Object {
            "danger": "#ff4136",
            "dark": "#484848",
            "disabled": "#999999",
            "info": "#75caeb",
            "light": "#f6f6f6",
            "primary": "#158cba",
            "success": "#28b62c",
            "warning": "#ff851b",
          },
          "social": Object {
            "facebook": "#3b5998",
            "google": "#dc4e41",
            "instagram": "#c32aa3",
            "linkedin": "#0077b5",
            "twitter": "#55acee",
            "vk": "#4c75a3",
          },
          "text": Object {
            "disabled": "#999999",
            "placeholder": "#e2e4e6",
            "primary": "#3d3d3d",
            "secondary": "#9e9e9e",
          },
        },
        "components": Object {},
        "radiuses": Object {
          "gigantic": 16,
          "huge": 12,
          "large": 8,
          "medium": 6,
          "small": 4,
          "tiny": 2,
        },
        "spacing": Object {
          "gigantic": 24,
          "huge": 20,
          "large": 16,
          "medium": 12,
          "small": 8,
          "tiny": 4,
        },
        "typography": Object {
          "fontFamily": "System",
          "fontSizes": Object {
            "large": 16,
            "medium": 14,
            "small": 12,
          },
          "fontWeights": Object {
            "bold": "900",
            "medium": "800",
            "normal": "600",
          },
          "iconSet": "MaterialIcons",
          "sizes": Object {
            "h1": Object {
              "fontSize": 31,
              "fontWeight": "900",
            },
            "h2": Object {
              "fontSize": 25,
              "fontWeight": "900",
            },
            "h3": Object {
              "fontSize": 19,
              "fontWeight": "800",
            },
            "h4": Object {
              "fontSize": 17,
              "fontWeight": "800",
            },
            "h5": Object {
              "fontSize": 16,
              "fontWeight": "800",
            },
            "h6": Object {
              "fontSize": 15,
              "fontWeight": "700",
            },
            "normal": Object {
              "fontSize": 14,
              "fontWeight": "700",
            },
            "small": Object {
              "fontSize": 12,
            },
          },
        },
      }
    }
  />
</View>
`;

exports[`Grid with Row renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "flexShrink": 1,
      },
      Object {
        "flexDirection": "column",
      },
    ]
  }
  styles={
    Object {
      "base": Object {
        "flexShrink": 1,
      },
      "directions": Object {
        "column": Object {
          "flexDirection": "column",
        },
        "row": Object {
          "flexDirection": "row",
        },
      },
    }
  }
  theme={
    Object {
      "colors": Object {
        "layout": Object {
          "backgroundColor": "#ffffff",
        },
        "palette": Object {
          "danger": "#ff4136",
          "dark": "#484848",
          "disabled": "#999999",
          "info": "#75caeb",
          "light": "#f6f6f6",
          "primary": "#158cba",
          "success": "#28b62c",
          "warning": "#ff851b",
        },
        "social": Object {
          "facebook": "#3b5998",
          "google": "#dc4e41",
          "instagram": "#c32aa3",
          "linkedin": "#0077b5",
          "twitter": "#55acee",
          "vk": "#4c75a3",
        },
        "text": Object {
          "disabled": "#999999",
          "placeholder": "#e2e4e6",
          "primary": "#3d3d3d",
          "secondary": "#9e9e9e",
        },
      },
      "components": Object {},
      "radiuses": Object {
        "gigantic": 16,
        "huge": 12,
        "large": 8,
        "medium": 6,
        "small": 4,
        "tiny": 2,
      },
      "spacing": Object {
        "gigantic": 24,
        "huge": 20,
        "large": 16,
        "medium": 12,
        "small": 8,
        "tiny": 4,
      },
      "typography": Object {
        "fontFamily": "System",
        "fontSizes": Object {
          "large": 16,
          "medium": 14,
          "small": 12,
        },
        "fontWeights": Object {
          "bold": "900",
          "medium": "800",
          "normal": "600",
        },
        "iconSet": "MaterialIcons",
        "sizes": Object {
          "h1": Object {
            "fontSize": 31,
            "fontWeight": "900",
          },
          "h2": Object {
            "fontSize": 25,
            "fontWeight": "900",
          },
          "h3": Object {
            "fontSize": 19,
            "fontWeight": "800",
          },
          "h4": Object {
            "fontSize": 17,
            "fontWeight": "800",
          },
          "h5": Object {
            "fontSize": 16,
            "fontWeight": "800",
          },
          "h6": Object {
            "fontSize": 15,
            "fontWeight": "700",
          },
          "normal": Object {
            "fontSize": 14,
            "fontWeight": "700",
          },
          "small": Object {
            "fontSize": 12,
          },
        },
      },
    }
  }
>
  <View
    size={null}
    style={
      Array [
        undefined,
        Object {
          "flexDirection": "row",
        },
        Object {
          "flex": 1,
        },
      ]
    }
    styles={
      Object {
        "base": Object {
          "flexDirection": "row",
        },
      }
    }
    theme={
      Object {
        "colors": Object {
          "layout": Object {
            "backgroundColor": "#ffffff",
          },
          "palette": Object {
            "danger": "#ff4136",
            "dark": "#484848",
            "disabled": "#999999",
            "info": "#75caeb",
            "light": "#f6f6f6",
            "primary": "#158cba",
            "success": "#28b62c",
            "warning": "#ff851b",
          },
          "social": Object {
            "facebook": "#3b5998",
            "google": "#dc4e41",
            "instagram": "#c32aa3",
            "linkedin": "#0077b5",
            "twitter": "#55acee",
            "vk": "#4c75a3",
          },
          "text": Object {
            "disabled": "#999999",
            "placeholder": "#e2e4e6",
            "primary": "#3d3d3d",
            "secondary": "#9e9e9e",
          },
        },
        "components": Object {},
        "radiuses": Object {
          "gigantic": 16,
          "huge": 12,
          "large": 8,
          "medium": 6,
          "small": 4,
          "tiny": 2,
        },
        "spacing": Object {
          "gigantic": 24,
          "huge": 20,
          "large": 16,
          "medium": 12,
          "small": 8,
          "tiny": 4,
        },
        "typography": Object {
          "fontFamily": "System",
          "fontSizes": Object {
            "large": 16,
            "medium": 14,
            "small": 12,
          },
          "fontWeights": Object {
            "bold": "900",
            "medium": "800",
            "normal": "600",
          },
          "iconSet": "MaterialIcons",
          "sizes": Object {
            "h1": Object {
              "fontSize": 31,
              "fontWeight": "900",
            },
            "h2": Object {
              "fontSize": 25,
              "fontWeight": "900",
            },
            "h3": Object {
              "fontSize": 19,
              "fontWeight": "800",
            },
            "h4": Object {
              "fontSize": 17,
              "fontWeight": "800",
            },
            "h5": Object {
              "fontSize": 16,
              "fontWeight": "800",
            },
            "h6": Object {
              "fontSize": 15,
              "fontWeight": "700",
            },
            "normal": Object {
              "fontSize": 14,
              "fontWeight": "700",
            },
            "small": Object {
              "fontSize": 12,
            },
          },
        },
      }
    }
  />
</View>
`;

exports[`Row renders correctly 1`] = `
<View
  size={null}
  style={
    Array [
      undefined,
      Object {
        "flexDirection": "row",
      },
      Object {
        "flex": 1,
      },
    ]
  }
  styles={
    Object {
      "base": Object {
        "flexDirection": "row",
      },
    }
  }
  theme={
    Object {
      "colors": Object {
        "layout": Object {
          "backgroundColor": "#ffffff",
        },
        "palette": Object {
          "danger": "#ff4136",
          "dark": "#484848",
          "disabled": "#999999",
          "info": "#75caeb",
          "light": "#f6f6f6",
          "primary": "#158cba",
          "success": "#28b62c",
          "warning": "#ff851b",
        },
        "social": Object {
          "facebook": "#3b5998",
          "google": "#dc4e41",
          "instagram": "#c32aa3",
          "linkedin": "#0077b5",
          "twitter": "#55acee",
          "vk": "#4c75a3",
        },
        "text": Object {
          "disabled": "#999999",
          "placeholder": "#e2e4e6",
          "primary": "#3d3d3d",
          "secondary": "#9e9e9e",
        },
      },
      "components": Object {},
      "radiuses": Object {
        "gigantic": 16,
        "huge": 12,
        "large": 8,
        "medium": 6,
        "small": 4,
        "tiny": 2,
      },
      "spacing": Object {
        "gigantic": 24,
        "huge": 20,
        "large": 16,
        "medium": 12,
        "small": 8,
        "tiny": 4,
      },
      "typography": Object {
        "fontFamily": "System",
        "fontSizes": Object {
          "large": 16,
          "medium": 14,
          "small": 12,
        },
        "fontWeights": Object {
          "bold": "900",
          "medium": "800",
          "normal": "600",
        },
        "iconSet": "MaterialIcons",
        "sizes": Object {
          "h1": Object {
            "fontSize": 31,
            "fontWeight": "900",
          },
          "h2": Object {
            "fontSize": 25,
            "fontWeight": "900",
          },
          "h3": Object {
            "fontSize": 19,
            "fontWeight": "800",
          },
          "h4": Object {
            "fontSize": 17,
            "fontWeight": "800",
          },
          "h5": Object {
            "fontSize": 16,
            "fontWeight": "800",
          },
          "h6": Object {
            "fontSize": 15,
            "fontWeight": "700",
          },
          "normal": Object {
            "fontSize": 14,
            "fontWeight": "700",
          },
          "small": Object {
            "fontSize": 12,
          },
        },
      },
    }
  }
/>
`;
